docs: ProgramNotificationTemplate in new Tracker module[DHIS2-17502] (#1425)
dhis2/dhis2-docs: src/developer/web-api/tracker.md

Each program notification template is transformed into either a MessageConversation object or a ProgramMessage object, depending on whether the recipient is external or internal. 
These intermediate objects will contain only the translated message and subject text.

There are several configuration parameters in the Program Notification Template that are essential for the proper functioning of notifications.
 These parameters are explained in the table below.

    POST /api/programNotificationTemplates

```json
{
	"name": "Case notification",
	"notificationTrigger": "ENROLLMENT",
	"subjectTemplate": "Case notification V{org_unit_name}",
	"displaySubjectTemplate": "Case notification V{org_unit_name}",
	"notifyUsersInHierarchyOnly": false,
	"sendRepeatable": false,
	"notificationRecipient": "ORGANISATION_UNIT_CONTACT",
	"notifyParentOrganisationUnitOnly": false,
	"displayMessageTemplate": "Case notification A{h5FuguPFF2j}",
	"messageTemplate": "Case notification A{h5FuguPFF2j}",
	"deliveryChannels": [
		"EMAIL"
	]
}
```

The fields are explained in the following table.

Table: Program Notification Template payload

| Field | Required | Description | Values |
|---|---|---|---|

> | name | Yes | name of Program Notification Tempalte | case-notification-alert |

Reviewer: ```suggestion
| name | Yes | name of Program Notification Template | case-notification-alert |
```

| notificationTrigger | Yes | When notification should be triggered. Possible values are ENROLLMENT, COMPLETION, PROGRAM_RULE, SCHEDULED_DAYS_DUE_DATE| ENROLLMENT |
| subjectTemplate | No | Subject template string | Case notification V{org_unit_name} |
| messageTemplate | Yes | Message template string | Case notification A{h5FuguPFF2j} |

> | notificationRecipient | YES | Who is going to receive notification. Possible values are USER_GROUP, ORGANISATION_UNIT_CONTACT, TRACKED_ENTITY_INSTANCE, USERS_AT_ORGANISATION_UNIT, DATA_ELEMENT, PROGRAM_ATTRIBUTE, WEB_HOOK  | USER_GROUP |

Reviewer: ```suggestion
| notificationRecipient | Yes | Who is going to receive notification. Possible values are USER_GROUP, ORGANISATION_UNIT_CONTACT, TRACKED_ENTITY_INSTANCE, USERS_AT_ORGANISATION_UNIT, DATA_ELEMENT, PROGRAM_ATTRIBUTE, WEB_HOOK  | USER_GROUP |
```

| deliveryChannels | No | Which channel should be used for this notification. It can be either SMS, EMAIL or HTTP | SMS |
| sendRepeatable | No | Whether notification should be sent multiple times | false |

> NOTE: WEB_HOOK notificationRecipient is used only to POST http request to an external system. Make sure to choose HTTP delivery channel when using WEB_HOOK.

Reviewer: ```suggestion
NOTE: The WEB_HOOK notificationRecipient is used exclusively for sending HTTP POST requests to external systems. Ensure that the HTTP delivery channel is selected when using this option.
```

      "trackedEntity": {
        "id": "UN810PwyVYO"
      },
      "organisationUnit": {
        "id": "Rp268JB6Ne4"
      },
      "phoneNumbers": [
        "55512345",
        "55545678"
      ],
      "emailAddresses": [
        "johndoe@mail.com",
        "markdoe@mail.com"
      ]
    },
    "enrollment": {
      "id": "f3rg8gFag8j"
    },
    "event": {
      "id": "pSllsjpfLH2"
    },
    "deliveryChannels": [
      "SMS", "EMAIL"
    ],
    "notificationTemplate": "Zp268JB6Ne5",
    "subject": "Outbreak alert",
    "text": "An outbreak has been detected",
    "storeCopy": false
  }]
}
```

The fields are explained in the following table.

Table: Program message payload

| Field | Required | Description | Values |
|---|---|---|---|
| recipients | Yes | Recipients of the program message. At least one recipient must be specified. Any number of recipients / types can be specified for a message. | Can be trackedEntity, organisationUnit, an array of phoneNumbers or an array of emailAddresses. |
| enrollment | Enrollment which ProgramMessage is attached to | Enrollment. | Enrollment ID. |
| event      | Event which ProgramMessage is attached to | Event. | Event ID. |

> | deliveryChannels | Yes | Array of delivery channels. | SMS &#124; EMAIL |

Reviewer: Why is the pipe different here? Or was it not pipe?

Author: i think it says 2 possible values` SMS & EMAIL`

| subject | No | The message subject. Not applicable for SMS delivery channel. | Text. |
| text | Yes | The message text. | Text. |

> | storeCopy | No | Whether to store a copy of the program message in DHIS2. | false (default) &#124; true |

Reviewer: Pipe? Why is the pipe different here? Or was it not pipe?

	"sendRepeatable": false,
	"notificationRecipient": "ORGANISATION_UNIT_CONTACT",
	"notifyParentOrganisationUnitOnly": false,
	"displayMessageTemplate": "Case notification A{h5FuguPFF2j}",
	"messageTemplate": "Case notification A{h5FuguPFF2j}",
	"deliveryChannels": [
		"EMAIL"
	]
}
```

#### Table: Program Notification Template payload

| Field | Required | Description | Values |
|---|---|---|---|
| name | Yes | Name of the Program Notification Template | case-notification-alert |
| notificationTrigger | Yes | When notification should be triggered. Possible values are ENROLLMENT, COMPLETION, PROGRAM_RULE, SCHEDULED_DAYS_DUE_DATE | ENROLLMENT |
| subjectTemplate | No | Subject template string | Case notification V{org_unit_name} |
| messageTemplate | Yes | Message template string | Case notification A{h5FuguPFF2j} |
| notificationRecipient | Yes | Who is going to receive notification. Possible values are USER_GROUP, ORGANISATION_UNIT_CONTACT, TRACKED_ENTITY_INSTANCE, USERS_AT_ORGANISATION_UNIT, DATA_ELEMENT, PROGRAM_ATTRIBUTE, WEB_HOOK | USER_GROUP |
| deliveryChannels | No | Which channel should be used for this notification. It can be either SMS, EMAIL, or HTTP | SMS |

> | sendRepeatable | No | Whether notification should be sent multiple times | false |

Reviewer 2: I don't know anything about this feature, so I'm kind of reading it as an end user.
In case `sendRepeatable` is enabled, how many times will it be sent? is there a limit?

Author: Good question, @muilpp . There is no fixed limit, and practically, setting one wouldn’t be feasible.

This is mostly used for repeatable programStages, such as contact screening for a household of a TB patient. Since the number of household contacts can vary significantly, it’s difficult to set an exact limit. However, we can define a reasonable threshold—say, 50 contacts—assuming no one has more than 50 family members. 🙂

**NOTE:** The `WEB_HOOK` notificationRecipient is used exclusively for sending HTTP POST requests to external systems. Ensure that the HTTP delivery channel is selected when using this option.

## Retrieving and Deleting Program Notification Template

As `ProgramNotificationTemplate` is a type of metadata, you can create, update, and delete it just like other metadata.

---

### Program Messages

The program message feature enables you to send messages to tracked entities, contact addresses associated with organizational units, phone numbers, and email addresses. Messages can be sent using the `messages` resource.

**POST /api/messages**

#### Sending program messages

> Program messages can be sent using two delivery channels:

Reviewer 2: are these delivery channels different to the ones listed [here](https://github.com/dhis2/dhis2-docs/pull/1425/files#diff-4b5dabc4706dd8ae8c4c71cc43b6c8d08fa520c861a63cae3908ac407f201fffR3502)?
if they are the same, should we include `HTTP` as well?

Author: No, `ProgramNotificationTemplate` and `ProgramMessage` are different.
A `ProgramNotificationTemplate` is converted into a `ProgramMessage` if its deliveryChannels include `SMS` or `EMAIL`.
This POST request is specifically used to create a `ProgramMessage`, which will be sent as an `SMS` or `EMAIL` based on the specified delivery channel.
